NIFI-14110 Support to limit content size in PackageFlowFile

EndzeitBegins opened the pull request:
Due to using a different API to retrieve the FlowFiles the behaviour when working with multiple queues is no longer unspecified.

I had an circular dependency problem when depending on `nifi-mock` from `nifi-utils`, which is why I use an anonymous implementation of `FlowFile` inside the tests instead of `MockFlowFile`.

<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-14110](https://issues.apache.org/jira/browse/NIFI-14110)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

joewitt commented on nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/FlowFileFilters.java:
```diff
@@ -43,23 +43,20 @@ public static FlowFileFilter newSizeBasedFilter(final double maxSize, final Data
                 }
 
-                if ((size + flowFile.getSize() > maxBytes) || (count + 1 > maxCount)) {
+                if (size > maxBytes || count > maxCount) {
```
In the previous code count and size were not changed for this case.  The case of reject and terminate is terminal and this FlowFileFilterResult will not be further used nor those values in anyway referenced?

EndzeitBegins replied:
I'm not sure I correctly understood your comment. Please clarify if the following doesn't address your point.

My goal was to simplify the implementation by unifying the three points of computation (inside the first `if` statement, inside the predicate of the second `if`, and before the last return) into a single point.
I added test cases for a few selected scenarios before adjusting the implementation to ensure that the behavior remains the same.

The JavaDoc states:
>  Returns a new {@link FlowFileFilter} that will pull FlowFiles until the maximum file size has been reached, or the maximum FlowFile Count was been reached ...

Therefore, once either the maximum count or accumulated size limit is reached, all subsequent FlowFiles should be rejected.

Upon closer inspection, the old implementation contained a minor bug related to this behavior. If a FlowFile exceeding the size limit was initially rejected (with `REJECT_AND_TERMINATE`), and the caller subsequently passed an empty FlowFile to the filter, the filter would incorrectly return `ACCEPT_AND_CONTINUE` even though the limit had already been reached. While this behavior is technically incorrect, it's primarily a consequence of the caller not adhering to the `FlowFileFilterResult` contract and thus not a critical bug.
I enhanced one of the test cases to demonstrate this behavior.

Both the old and new implementations are susceptible to potential integer overflow issues for both `count` and `size`. I'm not sure if that's something worth addressing though.

I'm content with leaving the implementation of `newSizeBasedFilter` unchanged if that's prefered. I encountered the code while working on the desired changes for `PackageFlowFile` and deemed it worth some tests / minor refactor.

mosermw commented:
I recommend modifying a MultiProcessorUseCase or creating another UseCase in order to make the documentation for combining the two Batch Size properties clear.  It's very important to be clear that flowfiles will not be delayed in the input queue  waiting for a batch size to be reached.  It's also very important to support packaging exactly 1 flowfile.

While this improvement appears to be worthwhile, we should be very careful with configuration creep on PackageFlowFile.  It's only justification for existence is to be easier to use than MergeContent for a specific use case.  Too many features would ruin that justification.

EndzeitBegins commented:
Thank you for the useful feedback @mosermw. I've adjusted the documentation of the UseCases to clarify the batching behaviour of the processor.

`PackageFlowFile` in combination with `UnpackContent` is a useful pair of processors to transfer FlowFiles between NiFi clusters where the more robust approach using remote process groups is not applicable, e.g. due to network restrictions.
Packaging more than one FlowFile can improve efficiency both in storage and transmission.
In my opinion, when the content size of the FlowFiles to transfer can vary largely, being able to apply a soft constraint on the package size can be helpful. 

Personally I do not intent to add other properties to the processor at the moment.

mosermw commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PackageFlowFile.java:
```diff
@@ -120,17 +122,34 @@
+            .description("Maximum combined content size of FlowFiles to package into one output FlowFile. " +
+                    "Note, that FlowFiles whose content exceeds this limit are packaged separately.")
+            .required(true)
+            .defaultValue("1 GB")
```
After thinking about this more, and considering that this property enforces a soft upper limit on packaged file size, a large value here won't affect anything if Maximum Batch Size = 1.  So no changes needed.

EndzeitBegins commented:
Thank you for your review @mosermw. 

Apologies for the delayed response. I've now addressed the feedback.

The pull request has been rebased onto the latest main commit, resolving all conflicts.
Each `PropertyDescriptor` is now on its own line.
The redundant batching benefits description has been removed, and a newline added to the test file's end.

I'd appreciate an review of the updated changes.

mosermw commented:
Sorry it took so long to review this @EndzeitBegins.  I tested this again and the functionality and documentation looks good.  Code looks good.
+1
